archive/archive.go

```diff
@@ -0,0 +1,265 @@
+		if err != nil {
+			return Application{}, err
+		} else if !exist {
+			return Application{}, fmt.Errorf("Info.plist does not exist at: %s", infoPlistPth)
+		}
+		infoPlist, err := plistutil.NewPlistDataFromFile(infoPlistPth)
+		if err != nil {
+			return Application{}, err
+		}
+		application.InfoPlist = infoPlist
+	}
+
+	{
+		provisioningProfilePth := filepath.Join(appOrAppexPth, "embedded.mobileprovision")
+		exist, err := pathutil.IsPathExists(provisioningProfilePth)
+		if err != nil {
+			return Application{}, err
+		} else if !exist {
+			return Application{}, fmt.Errorf("embedded.mobileprovision does not exist at: %s", provisioningProfilePth)
+		}
+		profile, err := profileutil.NewProvisioningProfileInfoFromFile(provisioningProfilePth)
+		if err != nil {
+			return Application{}, err
+		}
+		application.ProvisioningProfile = profile
+	}
+
+	{
+		entitlementsPth := filepath.Join(appOrAppexPth, "archived-expanded-entitlements.xcent")
+		exist, err := pathutil.IsPathExists(entitlementsPth)
+		if err != nil {
+			return Application{}, err
+		} else if exist {
+			entitlements, err := plistutil.NewPlistDataFromFile(entitlementsPth)
+			if err != nil {
+				return Application{}, err
+			}
+
+			application.Entitlements = entitlements
+		}
+	}
+
+	return application, nil
+}
+
+// Applications ...
+type Applications struct {
+	MainApplication  Application
+	WatchApplication *Application
+}
+
+// NewApplications ...
+func NewApplications(applicationsDir string) (Applications, error) {
+	mainApplication := Application{}
+	mainApplicationPth := ""
+	{
+		pattern := filepath.Join(applicationsDir, "*.app")
+		pths, err := filepath.Glob(pattern)
+		if err != nil {
+			return Applications{}, err
+		}
+
+		if len(pths) == 0 {
+			return Applications{}, fmt.Errorf("Failed to find main application using pattern: %s", pattern)
+		} else if len(pths) > 1 {
+			log.Warnf("Multiple main applications found")
+			for _, pth := range pths {
+				log.Warnf("- %s", pth)
+			}
+
+			mainApplicationPth = pths[0]
+			log.Warnf("Using first: %s", mainApplicationPth)
+		} else {
+			mainApplicationPth = pths[0]
+		}
+
+		mainApplication, err = NewApplication(mainApplicationPth)
+		if err != nil {
+			return Applications{}, err
+		}
+	}
+
+	plugins := []Application{}
+	{
+		pattern := filepath.Join(mainApplicationPth, "PlugIns/*.appex")
+		pths, err := filepath.Glob(pattern)
+		if err != nil {
+			return Applications{}, err
+		}
+		for _, pth := range pths {
+			plugin, err := NewApplication(pth)
+			if err != nil {
+				return Applications{}, err
+			}
+
+			plugins = append(plugins, plugin)
+		}
+		mainApplication.Plugins = plugins
+	}
+
+	var watchApplicationPtr *Application
+	watchApplicationPth := ""
+	{
+		pattern := filepath.Join(mainApplicationPth, "Watch/*.app")
+		pths, err := filepath.Glob(pattern)
+		if err != nil {
+			return Applications{}, err
+		}
+
+		if len(pths) > 1 {
+			log.Warnf("Multiple watch applications found")
```

Comment:
Instead of printing warning message in a package should we fail instantly if multiple watch *.app found? Does it matter if instead of an error message we've print warning messages?